### PR TITLE
Audition submission

### PR DESCRIPTION
**Pull request by hassanharoon86**

### CODE REVIEW

File: app/controllers/auditions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      redirect_to root_path
+    else
+      render 'new'
+    end
+  end
+
+  private
+
+  def audition_params
+    params.require(:audition).permit(:first_name, :last_name, :email, :artist_name, :hear_about, :additional_info, :other_source, links_attributes: [:id, :link, :_destroy], genres: [])
```
Comment (asad-imtiaz) [inline app/controllers/auditions_controller.rb:20]: make this multi-lined.

File: app/javascript/packs/audition_submission.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    if(this.value == 'Other'){
+      $('#other-field').fadeIn();
+    }
+    else{
+      $('#other-field').fadeOut();
+    }
+  });
+});
+
+$(document).on('cocoon:after-insert', function(){
```
Comment (asad-imtiaz) [inline app/javascript/packs/audition_submission.js:12]: keep `function() {}` spacing consistent.

File: app/models/audition.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  NAME_REGEX = /\A[a-zA-Z]+\z/
+
+  GENRES = ['Acapella', 'Acid', 'Jazz', 'Acoustic', 'Acid', 'Acid Punk', 'Alternative', 'Hip Hop', 'Ambient',
+    'Avantgarde', 'Bass', 'Blues', 'Cabaret', 'Celtic Chamber', 'Chanson', 'Chorus', 'Christian Rap', 'Cinematic', 'Classical',
+    'Classic Rock', 'Club', 'Comedy', 'Country', 'Cult' ]
+
+  SOURCES = ['Facebook', 'Instagram', 'Twitter', 'Other']
+
+  has_many :links, dependent: :destroy
+  accepts_nested_attributes_for :links, reject_if: :all_blank, allow_destroy: :true
```
Comment (asad-imtiaz) [inline app/models/audition.rb:12]: add empty line before

File: app/models/audition.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  validates :first_name, :last_name , length: { maximum: 30 },format: { with: NAME_REGEX, message: "No special characters allowed" }
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }
```
Comment (asad-imtiaz) [inline app/models/audition.rb:15]: you can use `Devise::email_regexp`

File: app/models/audition.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  validates :first_name, :last_name, :email, :artist_name, :links, :genres, :hear_about, presence: true
+
+  before_validation :remove_empty_genre
+
+  private
+
+  def remove_empty_genre
+    genres.reject! { |g| g.empty? }
```
Comment (asad-imtiaz) [inline app/models/audition.rb:23]: you can use short-hand syntax for this.
`genres.reject!(&:empty?)`

File: app/views/auditions/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  <%= f.fields_for :links do |link| %>
+    <%= render 'link_fields', f: link %>
+  <% end %>
+
+  <div class="mb-3 actions d-flex justify-content-end" id="add-link">
+    <%= link_to_add_association 'Add link', f, :links, class: 'btn btn-outline-secondary' %><br>
+  </div>
+
+  Select Genres:
```
Comment (asad-imtiaz) [inline app/views/auditions/_form.html.erb:33]: can you make this a label?

File: config/routes.rb [hunk 7]
```diff
@@ -1,8 +1,10 @@
 Rails.application.routes.draw do
   devise_for :users
-
   root to: 'home#index'
 
   resources :artists
   resources :managers
+  resources :auditions, only: [:new, :create]
+
```
Comment (asad-imtiaz) [inline config/routes.rb:8]: you can map this route inside `resources :auditions` block

File: app/views/auditions/_form.html.erb (hunk 6 above)
Comment (hassanharoon86) [reply app/views/auditions/_form.html.erb:33]: Yes I can

**Pull request merged**